gadi_jupyter

```diff
@@ -102,6 +104,43 @@ echo "qsub ${SUBMITOPTS}"
 # Kill the job if this top-level script is cancelled while the job is still in the queue
 trap "{ echo 'Stopping queued job... (Ctrl-C will leave job in the queue)' ; $SSH \"$LOGINNODE\" <<< \"qdel \\\$(cat \\$WORKDIR/jobid)\" ; }" EXIT
 
+if $RECONNECT; then
+
+message=$(
+$SSH -q "$LOGINNODE" <<EOF | tail -n 1
+PROJECT="$PROJECT"
+# Look for processes titled jupyter-notebook. 
+processes=(\$(/g/data/hh5/public/apps/nci_scripts/uqstat -f csv  | grep jupyter-notebook | grep \$PROJECT | sed 's/,/\n/g'))
+if [[ -n \$processes ]];
+then
+    set -eu
+    # Count number of jupyter-notebook sessions.
+    repeat=(\$( echo \${processes[@]} | tr " " "\n" | sort | uniq -c | sort -k2nr | grep jupyter-notebook | sed -e 's/^ *//g' | tr " " "\n" ))
+    # TO DO: Perhaps join all sessions or kill all processes. ATM, it will connect 
+    # to first session.
+    if [[ \$repeat != "1" ]]; 
+    then
+        echo "There are several sessions running"; 
+        qcat \${processes[0]} | head -n 1
+    else
+        # Extract process information
+        qcat \${processes[0]} | head -n 1
+    fi
+else
+    # If job doesn't exist then the message is overwritten
+    echo "New Job"
+fi
+EOF
+)
+else
+    message="New Job"
+fi
+
+if [[ ! $RECONNECT ]] || [[ $message == "New Job" ]]
+then
+
```

Comment:
Or better, do an early exit
```bash
# Check if resubmitting, message file exists and job is still running
if [ "$RESUBMIT" = "TRUE" -a -f \$WORKDIR/message ] && qstat \$(cat \$WORDIR/jobid) 2> /dev/null; then
    cat \$WORKDIR/message
    exit
fi

# Perform initial startup

mkdir -p "\$WORKDIR"
rm -f "\$WORKDIR/message"
...
```

Reply:
This implementation uses `qcat` instead of the data stored in `/scratch/$PROJECT/\$USER/tmpdir` to avoid the issue of the files been rewritten,  therefore it supports multiple connections, as the information of qcat is a temporal file unique to each job submission. Wouldn't it be better to use `qcat` instead of: 
```suggestion
qsub $SUBMITOPTS -l "storage=\$storage" -j oe -o "\$WORKDIR/pbs.log" "\$WORKDIR/runjp.sh" > "\$WORKDIR/jobid"
...
cat "\$WORKDIR/message"
```

Reply:
Multiple jupyter instances at once aren't currently a goal for the script - is there a need for that?  It would require more thought on how to handle things like cancelling the job while it's in the queue etc.

I guess either mechanism for the message will work, might be a bit slower through qcat if the queue is bogged down but that may not be a big impact.

What are the pipelines here doing? Note that $PROJECT is by default an empty string
```bash
processes=(\$(/g/data/hh5/public/apps/nci_scripts/uqstat -f csv  | grep jupyter-notebook | grep \$PROJECT | sed 's/,/\n/g'))

repeat=(\$( echo \${processes[@]} | tr " " "\n" | sort | uniq -c | sort -k2nr | grep jupyter-notebook | sed -e 's/^ *//g' | tr " " "\n" ))
```

Reply:
I some times use multiple instances when vdi is down. I presume its not a feature that many people will use. In fact, we should probably discourage people from doing it. 

If $PROJECT is defined, then it will only extract processes within the given project. If $PROJECT is an empty string, then grep will return the string containing the instances.